base/ltcounts.dtx

```diff
@@ -305,22 +307,56 @@
 %  \begin{macro}{\@definecounter}
 %  \changes{v1.1b}{1995/05/20}{Streamlined code}
 %  \changes{v1.1c}{1995/05/20}{And do it right}
+%  \changes{v1.1n}{2023/06/15}{define theHfoo (used for internal links)}
 %
 %    \begin{macrocode}
+%</2ekernel>
+%<latexrelease>\IncludeInRelease{2023/11/01}{\@definecounter}
+%<latexrelease>                             {provide theHfoo commands}%
+%<*2ekernel|latexrelease>
 \def\@definecounter#1{\expandafter\newcount\csname c@#1\endcsname
      \setcounter{#1}\z@
      \global\expandafter\let\csname cl@#1\endcsname\@empty
      \@addtoreset{#1}{@ckpt}%
      \global\expandafter\let\csname p@#1\endcsname\@empty
+     \expandafter\gdef\csname theH#1\endcsname{\the\value{#1}}%
      \expandafter
      \gdef\csname the#1\expandafter\endcsname\expandafter
           {\expandafter\@arabic\csname c@#1\endcsname}}
+%<latexrelease>\EndIncludeInRelease
+%</2ekernel|latexrelease>
+%<latexrelease>\IncludeInRelease{0000/00/00}{\@definecounter}
+%<latexrelease>                             {provide theHfoo commands}%%
+%<latexrelease>\def\@definecounter#1{\expandafter\newcount\csname c@#1\endcsname
+%<latexrelease>     \setcounter{#1}\z@
+%<latexrelease>     \global\expandafter\let\csname cl@#1\endcsname\@empty
+%<latexrelease>     \@addtoreset{#1}{@ckpt}%
+%<latexrelease>     \global\expandafter\let\csname p@#1\endcsname\@empty
+%<latexrelease>     \expandafter
+%<latexrelease>     \gdef\csname the#1\expandafter\endcsname\expandafter
+%<latexrelease>          {\expandafter\@arabic\csname c@#1\endcsname}}
+%<latexrelease>\EndIncludeInRelease
+%<*2ekernel>
 %    \end{macrocode}
 %  \end{macro}
 %
 % \begin{macro}{\@addtoreset}
+%  \changes{v1.1n}{2023/06/15}{add the parent theHfoo if a counter is reset}
 %    \begin{macrocode}
-\def\@addtoreset#1#2{\expandafter\@cons\csname cl@#2\endcsname {{#1}}}
+%</2ekernel>
+%<latexrelease>\IncludeInRelease{2023/11/01}{\@addtoreset}
+%<latexrelease>                             {provide theHfoo commands}%
+%<*2ekernel|latexrelease>
+\def\@addtoreset#1#2{\expandafter\@cons\csname cl@#2\endcsname {{#1}}%
+  \expandafter\gdef\csname theH#1\endcsname{\csname theH#2\endcsname.\the\value{#1}}%
```

Comment:
Does this work as intended in the following cases? 
--  Counter A is reset by Counter B and 
    Counter A is also reset by Counter C ;
--  Counter A is reset by Counter B and 
    Counter B is reset by Counter D .

Reply:
The exact definition of `\theHA` depends naturally on the order of the \@addtoreset calls. It "works" if one get unique `\theHA` value, which isn't 100% garantied -- users can always find a way to build something odd, -- but it is the definition used by hyperref since many years, so it is compatible with the current behaviour.